src/eytzinger_benchmark.zig

```diff
@@ -0,0 +1,319 @@
+    std.log.info("WT: Wall time/search", .{});
+    std.log.info("UT: utime time/search", .{});
+    std.log.info("CY: CPU cycles/search", .{});
+    std.log.info("IN: instructions/search", .{});
+    std.log.info("CR: cache references/search", .{});
+    std.log.info("CM: cache misses/search", .{});
+    std.log.info("BM: branch misses/search", .{});
+
+    var seed: u64 = undefined;
+    try std.os.getrandom(std.mem.asBytes(&seed));
+    var prng = std.rand.DefaultPrng.init(seed);
+
+    // Allocate from heap just once. All page allocations reuse this buffer.
+    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
+    defer arena.deinit();
+    const blob_size = GiB;
+    var blob = try arena.allocator.alloc(u8, blob_size);
+
+    inline for (kv_types) |kv| {
+        inline for (values_per_page) |values_count, v| {
+            inline for (summary_sizes[v]) |keys_count| {
+                try run_benchmark(.{
+                    .blob_size = blob_size,
+                    .key_size = kv.key_size,
+                    .value_size = kv.value_size,
+                    .keys_count = keys_count,
+                    .values_count = values_count,
+                    .searches = searches,
+                }, blob, &prng.random);
+            }
+        }
+    }
+}
+
+fn run_benchmark(comptime layout: Layout, blob: []u8, random: *std.rand.Random) !void {
+    assert(blob.len == layout.blob_size);
+    const Eytzinger = eytzinger(layout.keys_count - 1, layout.values_count);
+    const Val = Value(layout);
+    const Key = Val.Key;
+    const Page = struct {
+        summary: [layout.keys_count]Key,
+        values: [layout.values_count]Val,
+    };
+    const page_count = layout.blob_size / @sizeOf(Page);
+
+    // Search pages and keys in random order.
+    var page_picker = shuffled_index(page_count, random);
+    var value_picker = shuffled_index(layout.values_count, random);
+
+    // Generate 1GiB worth of 24KiB pages.
+    var blob_alloc = std.heap.FixedBufferAllocator.init(blob);
+    var pages = try blob_alloc.allocator.alloc(Page, page_count);
+    random.bytes(std.mem.sliceAsBytes(pages));
+    for (pages) |*page| {
+        for (page.values) |*value, i| value.key = i;
+        Eytzinger.layout_from_keys_or_values(
+            Key, Val, Val.key_from_value, Val.max_key,
+            &page.values, &page.summary);
+    }
+
+    const stdout = std.io.getStdOut().writer();
+    {
+        var benchmark = try Benchmark.begin();
+        var i: usize = 0;
+        var v: usize = 0;
+        while (i < layout.searches) : (i += 1) {
+            const page_index = page_picker[i % page_picker.len];
+            const target = value_picker[v % value_picker.len];
+            const page = pages[page_index];
+            const bounds = Eytzinger.search_values(Key, Val, Val.key_compare, &page.summary, &page.values, target);
+            assert(bounds.len != 0);
+            const hit = if (bounds.len == 1) bounds[0]
```

Comment:
I think it might help Eytzinger to just skip this branch and always do the binary search. Then the CPU doesn't need to stop and think. Although it's probably always going to be predicted by the branch predictor (since unless I'm mistaken we always search for hits), it's better to be explicit and not rely on this (golden rule: the less assumptions on the hardware the better), plus that's how we would do it within the LSM tree — go directly into binary search.

Reply:
I'd be curious if this has an impact.

Reply:
No impact that I can see, but the results are posted to the google sheet.